[IMPROVEMENT] Now this package using Laravel 5.3 notification

spyric opened the pull request:
**In raising this pull request, I confirm the following (please check boxes):**

- [x] I have read and understood the [contributors guide]().
- [x] I have checked that another pull request for this purpose does not exist.
- [x] I have considered, and confirmed that this submission will be valuable to others.
- [x] I have added tests to prove that the code in this PR works.

---

### Description
This PR for #11 .

Let me tell how it's works.

You can install new version and if your app suits next requirements:
* You have `\App\User` class
* You have added `scopeCanBeNotifiedAboutFailedJobs` method to `\App\User`.
You're ready to use it oit of the box,

You can publish config and get next abilities:
* Change Notifiable class
* Change Notification class
* Set channels
* Set criteria method's name for notifiables

Also you can add additional notifications.
For example what it you send very important email to your customer and if fails... Of course developers will get this error, but also we can setup ability to notify about this issue our sale merchant without any technical information, just to inform responsible person for client that something went wrong, so you can add to  config next lines:

```
'App\Jobs\SendImportantEmail' => [
    'notifiable' => \App\User::class,
    'notification' => \Spatie\FailedJobMonitor\Notification::class,
    'via' => ['mail'],
    //'filter' => 'canBeNotifiedAboutFailedJobs'
],
```

And `App\Jobs\SendImportantEmail` job will fail `*` and `App\Jobs\SendImportantEmail` will be used for notification.


So I think this solution is flexible:
* You're not attached to notifiable entity
* You can easily customize notification
* You can notify not only developers but users too about some important issues without any sensitive data.

All data (exception and payload) are sening as attachment via email.

I'm ready to hear your option, @freekmurze

freekmurze commented on README.md:
```diff
@@ -25,51 +25,68 @@ Next up, the service provider must be registered:
 'providers' => [
     ...
-    Spatie\FailedJobMonitor\FailedJobMonitorServiceProvider::class,
+    Spatie\FailedJobMonitor\ServiceProvider::class,
```
I prefer the old name. Could you rename the service provider back to `FailedJobMonitorServiceProvider`?

freekmurze commented on README.md:
```diff
@@ -25,51 +25,68 @@ Next up, the service provider must be registered:
+    '*' => [
+        'notifiable' => \App\User::class,
+        'notification' => \Spatie\FailedJobMonitor\Notification::class,
+        'via' => ['mail'],
```
Let's use the default Laravel names. Rename `via` to `channels`.

freekmurze commented on README.md:
```diff
@@ -25,51 +25,68 @@ Next up, the service provider must be registered:
+        'notifiable' => \App\User::class,
+        'notification' => \Spatie\FailedJobMonitor\Notification::class,
+        'via' => ['mail'],
+        //'filter' => 'canBeNotifiedAboutFailedJobs'
```
I don't like code comments in the config file. Remove this line.

freekmurze commented on resources/lang/en/mail.php:
```diff
@@ -0,0 +1,9 @@
+<?php
+
+return [
```
Very cool that you took the time to put all string in language files 👍

freekmurze commented on README.md:
````diff
@@ -25,51 +25,68 @@ Next up, the service provider must be registered:
 ```php
 return [
+    '*' => [
+        'notifiable' => \App\User::class,
````
I don't like the fact that de default notifiable has to be a user of the app.

 Replace this by a notifiable like the one in [laravel-backup](https://github.com/spatie/laravel-backup/blob/master/config/laravel-backup.php#L83)

freekmurze commented on README.md:
````diff
@@ -25,51 +25,68 @@ Next up, the service provider must be registered:
 
 ```php
 return [
+    '*' => [
````
I'd like to keep this package very simple. Adding job-specific config brings a lot of complexity with it. My thinking is that most users will want a notification for every job that fails.  Could you remove this functionality throughout this PR?

freekmurze commented on src/Notification.php:
```diff
@@ -0,0 +1,85 @@
+            });
+    }
+
+    protected function buildException(\Exception $e)
```
Generally I'm not a fan of abbreviations. Use `$exception` here.

freekmurze commented on src/Notification.php:
```diff
@@ -0,0 +1,85 @@
+    protected function buildException(\Exception $e)
+    {
+        $response = '';
+        while ($e !== null) {
```
Running over each exception is probably overkill. Just use `$exception->getTraceAsString()`. Here. It'll provide enough detail.

spyric commented:
I have updated code according your review. Now Travis failing with prefer-lowest config. I will look later today.

freekmurze commented on README.md:
```diff
@@ -36,40 +36,52 @@ Next, you must publish the config file:
+    'channels'   => ['mail', 'slack'],
+    'routes'     => [
+        'mail' => [
+            'to' => 'i.spyric@gmail.com',
```
Use the more generic `email@example.com` here.

freekmurze commented on README.md:
```diff
@@ -36,40 +36,52 @@ Next, you must publish the config file:
+        ],
+
+        'slack' => [
+            'webhook_url' => 'https://hooks.slack.com/services/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',
```
I hope this is not a real hook.  Just leave this one empty.

freekmurze commented on README.md:
````diff
@@ -36,40 +36,52 @@ Next, you must publish the config file:
+This is the contents of the default configuration file.  Most options are self-explanatory.
 
 ```php
 return [
````
to make it clear to the users what this config is for, add some comments.

You can use [these lines](https://github.com/spatie/laravel-backup/blob/master/config/laravel-backup.php#L79-L82) for the `laravel-backup` package.

freekmurze commented on composer.json:
```diff
@@ -28,13 +28,16 @@
-        "orchestra/testbench":"^3.3",
-        "maknz/slack": "^1.5"
-
+        "orchestra/testbench":"^3.3.1"
```
use `~3.3.1` we don't want to pull in `3.4.x`.

freekmurze commented on config/laravel-failed-job-monitor.php:
```diff
@@ -1,26 +1,16 @@
+    'channels'     => ['mail', 'slack'],
+    'routes'       => [
+        'mail' => [
+            'to' => 'your@mail',
```
use `email@example.com` here

freekmurze commented on config/laravel-failed-job-monitor.php:
```diff
@@ -1,26 +1,16 @@
 <?php
 
 return [
+    'notifiable'   => \Spatie\FailedJobMonitor\Notifiable::class,
```
Apply all comments that I give at the config file in the readme here as well.

freekmurze commented on src/FailedJobMonitorServiceProvider.php:
```diff
@@ -2,9 +2,9 @@
 namespace Spatie\FailedJobMonitor;
 
-use Illuminate\Support\ServiceProvider;
+use Illuminate\Support\ServiceProvider as ServiceProviderBase;
```
Small nitpick: I tend to use `IlluminateServiceProvider` when I override an SP. Could you use that name as well? That's more consistent with the rest of our packages.

freekmurze commented on src/FailedJobNotifier.php:
```diff
@@ -2,25 +2,30 @@
+    public function register()
     {
-        $sender = $this->getChannelInstance($sender);
+        app(QueueManager::class)->failing(function (JobFailed $event) {
```
Cool, this is much better than in the previous review! 👍

freekmurze commented on tests/Dummy/TestQueueManager.php:
```diff
@@ -0,0 +1,80 @@
+            $payload = json_encode($this->createPlainPayload($job, $data));
+        }
+
+        if (JSON_ERROR_NONE !== json_last_error()) {
```
To me this is Yoda-style 😄 
Could you switch the arguments around?

spyric replied:
If fact createPayload took it from laravel/framework :D https://github.com/laravel/framework/blob/5.3/src/Illuminate/Queue/Queue.php#L79 I will rewrite it for you.

freekmurze commented on tests/Dummy/TestQueueManager.php:
```diff
@@ -0,0 +1,80 @@
+                    'commandName' => get_class($job),
+                    'command' => serialize(clone $job),
+                ],
+            ]);
```
I really dislike the use of `else` in general. Could you early return the `if` branch?

freekmurze commented:
Let me know when you've fixed the tests. Then I'll take a look again.

spyric commented:
@freekmurze . I have just added new commits with fixing according to your review, also added `channel`,`username`, `icon` to slack config. You can check this version. Tests now passes.

freekmurze commented:
Cool, thank you very much for your work on this, appreciate it a lot.

I'll pull in your changes to the `v2` branch and will make further review and make some changes myself.

spyric commented:
Thank you too. I got much experience while solving this task.
